File: lib/vcf-parser.ts
import { Readable } from "stream";
import { createInterface } from "readline";

export interface VcfVariant {
  chrom: string;
  pos: number;
  ref: string;
  alt: string;
  qual: number | null;
  filter: string | null;
  gene: string | null;
  consequence: string | null;
  hgvs_c: string | null;
  hgvs_p: string | null;
  gnomad_af: number | null;
  clinvar_sig: string | null;
  revel_score: number | null;
  spliceai_max: number | null;
  info_json: Record<string, unknown>;
}

export interface VcfMeta {
  pipeline_key: string | null;
  header_lines: string[];
}

/** Parse the INFO field into a key→value map */
function parseInfo(infoStr: string): Record<string, string | boolean> {
  if (!infoStr || infoStr === ".") return {};
  const result: Record<string, string | boolean> = {};
  for (const token of infoStr.split(";")) {
    const eq = token.indexOf("=");
    if (eq === -1) {
      result[token] = true;
    } else {
      result[token.slice(0, eq)] = token.slice(eq + 1);
    }
  }
  return result;
}

/** Extract VEP CSQ annotations from INFO map for a specific ALT allele */
function extractVepAnnotations(
  info: Record<string, string | boolean>,
  csqHeader: string[] | null,
  altAllele: string
): {
  gene: string | null;
  consequence: string | null;
  hgvs_c: string | null;
  hgvs_p: string | null;
  gnomad_af: number | null;
  clinvar_sig: string | null;
  revel_score: number | null;
  spliceai_max: number | null;
} {
  const empty = {
    gene: null,
    consequence: null,
    hgvs_c: null,
    hgvs_p: null,
    gnomad_af: null,
    clinvar_sig: null,
    revel_score: null,
    spliceai_max: null,
  };

  if (csqHeader && info["CSQ"]) {
    const csqValue = info["CSQ"] as string;
    // VEP may have multiple transcripts and alleles — filter by this ALT first
    const entries = csqValue.split(",");
    const alleleIdx = csqHeader.indexOf("Allele");

    // Filter to entries matching this ALT allele (if Allele field exists)
    const alleleMatched =
      alleleIdx >= 0
        ? entries.filter((e) => e.split("|")[alleleIdx] === altAllele)
        : entries;

    // Use allele-matched pool, or fall back to all entries if no match
    const pool = alleleMatched.length > 0 ? alleleMatched : entries;

    // Prefer canonical transcript (CANONICAL=YES)
    const canonIdx = csqHeader.indexOf("CANONICAL");
    const preferred =
      canonIdx >= 0
        ? pool.find((e) => e.split("|")[canonIdx] === "YES") ?? pool[0]
        : pool[0];

    const fields = preferred.split("|");
    const get = (name: string): string => {
      const i = csqHeader.indexOf(name);
      return i >= 0 ? (fields[i] ?? "") : "";
    };

    const gnomadRaw =
      get("gnomADe_AF") ||
      get("gnomAD_AF") ||
      get("gnomADg_AF") ||
      get("MAX_AF");
    const revelRaw = get("REVEL") || get("REVEL_score");

    // SpliceAI: max of DS_AG, DS_AL, DS_DG, DS_DL
    const saiScores = [
      parseFloat(get("SpliceAI_pred_DS_AG")),
      parseFloat(get("SpliceAI_pred_DS_AL")),
      parseFloat(get("SpliceAI_pred_DS_DG")),
      parseFloat(get("SpliceAI_pred_DS_DL")),
    ].filter((v) => !isNaN(v));
    const spliceaiMax = saiScores.length > 0 ? Math.max(...saiScores) : null;

    return {
      gene: get("SYMBOL") || get("Gene") || null,
      consequence: get("Consequence") || null,
      hgvs_c: get("HGVSc") || null,
      hgvs_p: get("HGVSp") || null,
      gnomad_af: gnomadRaw ? parseFloat(gnomadRaw) : null,
      clinvar_sig: get("CLIN_SIG") || get("ClinVar_CLNSIG") || null,
      revel_score: revelRaw ? parseFloat(revelRaw) : null,
      spliceai_max: spliceaiMax,
    };
  }

  // SnpEff ANN field fallback
  if (info["ANN"]) {
    const annValue = info["ANN"] as string;
    const first = annValue.split(",")[0].split("|");
    // ANN format: Allele|Annotation|Impact|Gene_Name|Gene_ID|...
    return {
      ...empty,
      gene: first[3] ?? null,
      consequence: first[1] ?? null,
    };
  }

  return empty;
}

/** Detect pipeline from VCF meta-header lines */
function detectPipeline(headerLines: string[]): string | null {
  const sourceLines = headerLines
    .filter((l) => l.startsWith("##source") || l.startsWith("##pipeline"))
    .join(" ")
    .toLowerCase();
  if (sourceLines.includes("dragen")) return "dragen_germline";
  if (sourceLines.includes("haplotypecaller")) return "gatk_haplotypecaller";
  if (sourceLines.includes("mutect2")) return "mutect2";
  if (sourceLines.includes("strelka")) return "strelka2";
  return null;
}

/** Parse VCF from a Node.js Readable stream.
 *  Calls onVariant for each parsed variant row.
 *  Returns VcfMeta after the stream ends.
 */
export async function parseVcf(
  stream: Readable,
  onVariant: (v: VcfVariant) => Promise<void>
): Promise<VcfMeta> {
  const rl = createInterface({ input: stream, crlfDelay: Infinity });

  const headerLines: string[] = [];
  let csqHeader: string[] | null = null;
  let lineNum = 0;

  for await (const line of rl) {
    lineNum++;
    if (line.startsWith("##")) {
      headerLines.push(line);
      // Parse VEP CSQ format header: ##INFO=<ID=CSQ,...,Format: A|B|C>
      if (line.includes("ID=CSQ") && line.includes("Format:")) {
        const fmt = line.split("Format:")[1]?.replace(/[">]/g, "").trim();
        if (fmt) csqHeader = fmt.split("|");
      }
      continue;
    }
    if (line.startsWith("#CHROM")) continue; // column header

    const cols = line.split("\t");
    if (cols.length < 8) continue;

    const [chrom, posStr, , ref, altField, qualStr, filterStr, infoStr] = cols;
    const qual = qualStr === "." ? null : parseFloat(qualStr);
    const filter = filterStr === "." ? null : filterStr;
    const info = parseInfo(infoStr);

    // Handle multi-allelic: split ALT on comma
    const alts = altField.split(",");
    for (const alt of alts) {
      if (!alt || alt === "*") continue; // skip spanning deletions

      const annotations = extractVepAnnotations(info, csqHeader, alt);
      const variant: VcfVariant = {
        chrom,
        pos: parseInt(posStr, 10),
        ref,
        alt,
        qual,
        filter,
        ...annotations,
        info_json: info as Record<string, unknown>,
      };
      await onVariant(variant);
    }
  }

  const pipeline_key = detectPipeline(headerLines);
  return { pipeline_key, header_lines: headerLines };
}

File: backend/app/lib/vcf_parser.py
"""VCF / BCF parser backed by cyvcf2 (htslib).

Parses a VCF or BCF file, extracts per-variant annotation from VEP CSQ
or flat ``CSQ_*`` INFO fields, detects the originating sequencing
pipeline from ``##source`` / ``##pipeline`` header lines, and invokes an
optional callback for each parsed variant.

Note on INFO field extraction
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
cyvcf2 ``v.INFO.keys()`` only enumerates fields declared in the VCF
header; it silently omits flat ``CSQ_*`` keys that East Genomics
pipelines emit without header declarations.  This module works around
the limitation by parsing the raw INFO column string directly.

Primary entry point
-------------------
parse_vcf(path, on_variant)
    Parse *path* and call *on_variant* for each emitted ``VcfVariant``.
    Returns a ``VcfMeta`` containing the detected pipeline key and raw
    header lines.
"""

from __future__ import annotations

import logging
from collections.abc import Callable
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any

import cyvcf2

from app.lib.pipeline_config import detect_pipeline_key

logger = logging.getLogger(__name__)


@dataclass
class VcfVariant:
    """Parsed representation of a single ALT allele from a VCF data line.

    Multi-allelic sites are split so that each ``VcfVariant`` carries
    exactly one ALT allele.  Spanning deletions (``*``) are skipped.
    Annotation fields are populated from VEP CSQ or flat ``CSQ_*`` INFO
    fields; fields absent from the annotation are ``None``.
    """
    chrom: str
    pos: int
    ref: str
    alt: str
    qual: float | None
    filter: str | None
    gene: str | None
    consequence: str | None
    hgvs_c: str | None
    hgvs_p: str | None
    gnomad_af: float | None
    clinvar_sig: str | None
    revel_score: float | None
    spliceai_max: float | None
    info_json: dict[str, Any] = field(default_factory=dict)


@dataclass
class VcfMeta:
    """Metadata produced after a complete VCF file has been parsed."""

    pipeline_key: str | None
    header_lines: list[str]


def _parse_info(info_str: str) -> dict[str, str | bool]:
    """Parse a VCF INFO column string into a key-value dict.

    Flag fields (no ``=`` sign) map to ``True``.  Returns an empty dict
    for ``"."`` or an empty string.
    """
    if not info_str or info_str == ".":
        return {}
    result: dict[str, str | bool] = {}
    for token in info_str.split(";"):
        eq = token.find("=")
        if eq == -1:
            result[token] = True
        else:
            result[token[:eq]] = token[eq + 1:]
    return result


_SPLICE_KEYS = [
    "SpliceAI_pred_DS_AG", "SpliceAI_pred_DS_AL",
    "SpliceAI_pred_DS_DG", "SpliceAI_pred_DS_DL",
]
_FLAT_SPLICE_KEYS = [f"CSQ_{k}" for k in _SPLICE_KEYS]


def _csq_field(entry: str, idx: int) -> str:
    """Safely return a pipe-separated field from a CSQ entry; returns '' if out of range."""
    parts = entry.split("|")
    return parts[idx] if 0 <= idx < len(parts) else ""


def _spliceai_max(scores: list[str]) -> float | None:
    """Return the maximum finite SpliceAI delta score from *scores*, or ``None``.

    Skips empty strings, ``"."`` placeholders, and IEEE 754 NaN values.
    """
    vals = []
    for s in scores:
        if s and s not in (".", ""):
            try:
                v = float(s)
                if v == v:  # NaN != NaN in IEEE 754 — intentionally filters NaN SpliceAI scores
                    vals.append(v)
            except ValueError:
                pass
    return max(vals) if vals else None


def _try_float(s: str) -> float | None:
    """Return *s* parsed as a float, or ``None`` if conversion fails."""
    try:
        return float(s) if s else None
    except (ValueError, TypeError):
        return None


def _extract_vep(info: dict, csq_header: list[str], alt: str) -> dict:
    """Extract annotation fields from a VEP CSQ INFO value.

    Selects the most appropriate CSQ entry for *alt*: first filters to
    allele-matching entries, then prefers the canonical transcript
    (``CANONICAL=YES``), falling back to the first entry overall.

    Args:
        info: Parsed INFO dict containing a ``"CSQ"`` key.
        csq_header: Ordered field names from the ``##INFO=<ID=CSQ,...Format:>``
            header line.
        alt: The ALT allele string to match against the ``Allele`` field.

    Returns:
        A dict with keys ``gene``, ``consequence``, ``hgvs_c``,
        ``hgvs_p``, ``gnomad_af``, ``clinvar_sig``, ``revel_score``,
        ``spliceai_max``, all typed as ``float | str | None``.
    """
    empty: dict = {k: None for k in [
        "gene", "consequence", "hgvs_c", "hgvs_p",
        "gnomad_af", "clinvar_sig", "revel_score", "spliceai_max",
    ]}
    csq_str = info.get("CSQ")
    if not isinstance(csq_str, str):
        return empty

    entries = csq_str.split(",")
    allele_idx = csq_header.index("Allele")   if "Allele"    in csq_header else -1
    canon_idx  = csq_header.index("CANONICAL") if "CANONICAL" in csq_header else -1

    matched = [e for e in entries if allele_idx >= 0 and _csq_field(e, allele_idx) == alt]
    pool = matched if matched else entries
    preferred = next(
        (e for e in pool if canon_idx >= 0 and _csq_field(e, canon_idx) == "YES"),
        pool[0],
    )

    flds = preferred.split("|")

    def get(name: str) -> str:
        i = csq_header.index(name) if name in csq_header else -1
        return flds[i] if 0 <= i < len(flds) else ""

    gnomad_raw = get("gnomADe_AF") or get("gnomAD_AF") or get("gnomADg_AF") or get("MAX_AF")
    revel_raw  = get("REVEL") or get("REVEL_score")
    sai_vals   = [get(k) for k in _SPLICE_KEYS]

    return {
        "gene":        get("SYMBOL") or get("Gene") or None,
        "consequence": get("Consequence") or None,
        "hgvs_c":      get("HGVSc") or None,
        "hgvs_p":      get("HGVSp") or None,
        "gnomad_af":   _try_float(gnomad_raw),
        "clinvar_sig": get("CLIN_SIG") or get("ClinVar_CLNSIG") or None,
        "revel_score": _try_float(revel_raw),
        "spliceai_max": _spliceai_max(sai_vals),
    }


def _extract_flat_csq(info: dict) -> dict:
    """Extract annotation fields from East Genomics flat ``CSQ_*`` INFO keys.

    Returns the same dict shape as ``_extract_vep()``.
    """
    def get(name: str) -> str:
        v = info.get(name, "")
        return v if isinstance(v, str) else ""

    gnomad_raw = get("CSQ_gnomADe_AF") or get("CSQ_gnomADg_AF")
    sai_vals   = [get(k) for k in _FLAT_SPLICE_KEYS]

    return {
        "gene":        get("CSQ_SYMBOL") or None,
        "consequence": get("CSQ_Consequence") or None,
        "hgvs_c":      get("CSQ_HGVSc") or None,
        "hgvs_p":      get("CSQ_HGVSp") or None,
        "gnomad_af":   _try_float(gnomad_raw),
        "clinvar_sig": get("CSQ_ClinVar_CLNSIG") or None,
        "revel_score": _try_float(get("CSQ_REVEL")),
        "spliceai_max": _spliceai_max(sai_vals),
    }


def parse_vcf(
    path: str | Path,
    on_variant: Callable[[VcfVariant], None] | None = None,
) -> VcfMeta:
    """Parse a VCF or BCF file using cyvcf2 and emit one variant per ALT allele.

    Uses cyvcf2 (htslib) for file I/O.  INFO fields are extracted from
    the raw VCF line string rather than through ``v.INFO.keys()`` because
    the cyvcf2 API silently omits undeclared INFO keys such as the flat
    ``CSQ_*`` fields produced by East Genomics pipelines.

    Args:
        path: Path to a VCF or BCF file.  BGZF-compressed files and BCF
            are supported natively by htslib.
        on_variant: Optional callback invoked for each parsed variant.
            Callers that want to collect variants should pass
            ``variants.append``.

    Returns:
        A ``VcfMeta`` containing the detected pipeline key (or ``None``
        if the header does not match any known pattern) and all
        ``##``-prefixed header lines.
    """
    vcf = cyvcf2.VCF(str(path))

    # Extract header lines for pipeline detection
    header_lines = [
        line for line in str(vcf.raw_header).splitlines()
        if line.startswith("##")
    ]

    # Parse CSQ FORMAT from header
    csq_header: list[str] | None = None
    for line in header_lines:
        if "ID=CSQ" in line and "Format:" in line:
            fmt = line.split("Format:")[1].replace('"', "").replace(">", "").strip()
            csq_header = fmt.split("|")
            break

    for v in vcf:
        qual: float | None = v.QUAL  # cyvcf2 returns None if "."

        # Build info dict from raw INFO string (reliable for declared AND
        # undeclared fields such as flat CSQ_* fields from East Genomics pipelines).
        # cyvcf2 v.INFO.keys() only enumerates header-declared keys, so we
        # parse the raw VCF line for complete coverage.
        raw_cols = str(v).rstrip("\n").split("\t")
        filter_raw = raw_cols[6] if len(raw_cols) > 6 else "."
        filter_val: str | None = None if filter_raw in (".", "") else filter_raw
        info_str = raw_cols[7] if len(raw_cols) > 7 else "."
        info: dict[str, Any] = _parse_info(info_str)

        for alt in v.ALT:
            if not alt or alt == "*":
                continue

            if csq_header is not None and "CSQ" in info:
                annotations = _extract_vep(info, csq_header, alt)
            elif "CSQ_SYMBOL" in info or "CSQ_Consequence" in info:
                annotations = _extract_flat_csq(info)
            else:
                annotations = {k: None for k in [
                    "gene", "consequence", "hgvs_c", "hgvs_p",
                    "gnomad_af", "clinvar_sig", "revel_score", "spliceai_max",
                ]}

            variant = VcfVariant(
                chrom=v.CHROM,
                pos=v.POS,
                ref=v.REF,
                alt=alt,
                qual=qual,
                filter=filter_val,
                info_json=dict(info),
                **annotations,
            )
            if on_variant:
                on_variant(variant)

    vcf.close()
    pipeline_key = detect_pipeline_key(header_lines)
    return VcfMeta(pipeline_key=pipeline_key, header_lines=header_lines)
